IdentityCore/src/validation/MSIDAuthority.h

```diff
@@ -17,57 +15,75 @@
 //
 // THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 // IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 // AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
-//
-//------------------------------------------------------------------------------
 
 #import <Foundation/Foundation.h>
 #import "MSIDAuthorityResolving.h"
+#import "MSIDCache.h"
+
+extern NSString * _Nonnull const MSIDTrustedAuthority;
+extern NSString * _Nonnull const MSIDTrustedAuthorityWorldWide;
 
 @class MSIDOpenIdProviderMetadata;
 
 typedef void(^MSIDOpenIdConfigurationInfoBlock)(MSIDOpenIdProviderMetadata * _Nullable metadata, NSError * _Nullable error);
 
-extern NSString * _Nonnull const MSIDTrustedAuthorityWorldWide;
+@interface MSIDAuthority : NSObject <NSCopying>
+{
+@protected
+    NSURL *_url;
+    NSURL *_openIdConfigurationEndpoint;
+}
+
+@property (class, readonly, nonnull) MSIDCache *openIdConfigurationCache;
+
+@property (nonatomic, readonly, nonnull) NSURL *url;
+
+@property (nonatomic, readonly, nullable) NSURL *openIdConfigurationEndpoint;
+
+@property (nonatomic, readonly, nullable) MSIDOpenIdProviderMetadata *metadata;
+
+- (instancetype _Nullable )init NS_UNAVAILABLE;
++ (instancetype _Nullable )new NS_UNAVAILABLE;
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError * _Nullable __autoreleasing * _Nullable)error NS_DESIGNATED_INITIALIZER;
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                           rawTenant:(nullable NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError * _Nullable __autoreleasing * _Nullable)error;
+
+- (void)resolveAndValidate:(BOOL)validate
+         userPrincipalName:(nullable NSString *)upn
+                   context:(nullable id<MSIDRequestContext>)context
+           completionBlock:(nonnull MSIDAuthorityInfoBlock)completionBlock;
 
-@interface MSIDAuthority : NSObject
+- (nonnull NSURL *)networkUrlWithContext:(nullable id<MSIDRequestContext>)context;
 
-@property (class, readonly, nonnull) NSCache *openIdConfigurationCache;
+- (nonnull NSURL *)cacheUrlWithContext:(nullable id<MSIDRequestContext>)context;
 
-+ (BOOL)isADFSInstance:(nonnull NSString *)endpoint;
-+ (BOOL)isADFSInstanceURL:(nonnull NSURL *)endpointUrl;
-+ (BOOL)isConsumerInstanceURL:(nonnull NSURL *)authorityURL;
-+ (BOOL)isB2CInstanceURL:(nonnull NSURL *)endpointUrl;
+- (nonnull NSArray<NSURL *> *)cacheAliases;
```

Comment:
In the new cache we don't operate based on alias URLs, we operate based on environments which are strings. How can we now get a list of environment aliases?

Reply:
We still can get them from MSIDOauth2Factory. On the other hand, I don't think that MSIDOauth2Factory is a right place for such logic, that is why I suggest to revise it latter, see #222.